core/primitives/debug-derive/src/impls.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use quote::quote;
+use proc_macro2::TokenStream;
+use syn::{Ident, Data, DeriveInput, parse_quote};
+
+fn ident2str(name: &Ident) -> String {
+	format!("{}", name)
+}
+
+pub fn debug_derive(ast: &DeriveInput) -> proc_macro::TokenStream {
+	let name_str = ident2str(&ast.ident);
+	let implementation = implementation::derive(&name_str, &ast.data);
+	let name = &ast.ident;
+	let mut generics = ast.generics.clone();
+	let (impl_generics, ty_generics, where_clause) = {
+		let wh = generics.make_where_clause();
+		for t in ast.generics.type_params() {
+			let name = &t.ident;
+			wh.predicates.push(parse_quote!{ #name : core::fmt::Debug });
+		}
+		generics.split_for_impl()
+	};
+	let gen = quote!{
+		impl #impl_generics core::fmt::Debug for #name #ty_generics #where_clause {
+			fn fmt(&self, fmt: &mut core::fmt::Formatter) -> core::fmt::Result {
+				#implementation
+			}
+		}
+	};
+
+	gen.into()
+}
+
+#[cfg(not(any(test, feature = "std")))]
```

Comment:
Did you thought of using `CARGO_CFG_STD`?

Reply:
What does it do? I've actually changed the condition a bit, because `test` is unnecessary (we run tests in `std` anyway).

Reply:
I mean, instead of having the feature exposed by the derive macro, just check if the `CARGO_CFG_STD` is set while compiling a crate. Cargo sets environment variables for each cfg that is enabled for a crate and as the proc-macro is running as part of the build process, you can inspect this environment variables.

Reply:
Seems that the env variable is not set when compiling tests and I also find this approach more fragile than the usual feature-based conditional compilation.